drivers/scsi/ses.c

```diff
@@ -650,11 +663,20 @@ static int ses_intf_add(struct device *cdev,
 
 	if (!scsi_device_enclosure(sdev)) {
 		/* not an enclosure, but might be in one */
-		struct enclosure_device *prev = NULL;
+		if (scsi_is_ata(sdev)) {
+			struct ata_port *ap = ata_shost_to_port(sdev->host);
+			struct ahci_host_priv *hpriv = ap->host->private_data;
+			struct Scsi_Host *shost = hpriv->em_shost;
```

Comment:
This looks like a gross hack...

Reply:
`struct ata_host` has an array of `struct ata_port *`, and each port has a corresponding `struct Scsi_Host`. An AHCI controller is a specialization of `struct ata_host` with fields specific to AHCI (including several related to enclosure management) accessed through the host `private_data` field. The enclosure management features of an AHCI controller do not have a corresponding `struct ata_port`, so I have added the field `struct Scsi_Host *em_shost` to the AHCI struct for keeping track of the `struct Scsi_Host` allocated for SES emulation. That way we can go from a `struct scsi_device *sdev` for a device attached to an AHCI controller to the `struct Scsi_Host *shost` for its enclosure management.

What's gross about it? I could maybe clean it up a bit by hiding the details in a libahci function, and I do make the assumption that the ATA port belongs to an AHCI controller which is more of an oversight than a gross hack. I can add that check.

Reply:
I see it as huge layering violation for generic SES code to have access to the internal structures of the ATA driver.  I'd prefer it to not even know about the driver used.  My FreeBSD code is maximally opaque, using only numeric bus/target numbers for disk/slot mapping, that if needed could be reused for any other driver without adding more specific conditions, public headers, etc.

Reply:
Ok I can think about how to improve that, but for now is it ok to merge for nightly builds?